File: Sources/LedgelingsCore/Banter.swift
import Foundation

/// Who a creature is when it opens its mouth.
public struct Character: Codable, Equatable, Sendable {
    public var name: String
    public var persona: String

    public init(name: String, persona: String) {
        self.name = name
        self.persona = persona
    }
}

/// The words that go to the model. Pure string work, so it is testable without
/// a server: templates with `{placeholders}`, filled from a small dictionary.
public enum Banter {
    public static let placeholders = ["speaker", "speakerKind", "speakerPersona", "listener", "listenerKind", "listenerPersona", "situation", "line"]

    /// What the built-in creature is, for the prompt.
    public static let defaultKind = "a small square creature"

    public static let defaultSystemPrompt = """
    You are {speaker}, {speakerKind}, living on the edge of a computer screen. {speakerPersona}
    You are talking to {listener}, {listenerKind}, who lives on the same edge. {listenerPersona}
    Say ONE line to {listener}: a joke, a jab or a tease, at most 20 words, in your own voice.
    Output only the line. No quotes, no name prefix, no explanation.
    """

    public static let defaultLinePrompt = """
    Right now: {situation}
    Say your line to {listener}.
    """

    public static let defaultReplyPrompt = """
    Right now: {situation}
    {listener} just said to you: "{line}"
    Answer back in ONE line, in character, at most 20 words.
    """

    public static let defaultCharacters: [Character] = [
        Character(name: "Blocky", persona: "Grumpy and proud. Hates the mouse cursor. Thinks the bottom edge is the only respectable edge."),
        Character(name: "Pip", persona: "Cheerful and easily impressed. Loves the ceiling. Laughs at everything, including insults."),
        Character(name: "Mortimer", persona: "Old and philosophical. Speaks slowly, quotes wisdom he made up, sighs a lot."),
        Character(name: "Zed", persona: "Sleepy. Would rather be napping. Every sentence drifts toward bed."),
        Character(name: "Dot", persona: "Tiny, fast and sarcastic. Brags about speed. Calls everyone else a boulder."),
        Character(name: "Ruth", persona: "Bossy, organised, keeps count of everything. Disapproves of jumping."),
    ]

    /// How long a bubble stays up for a line of typical length, in seconds.
    public static let defaultBubbleSeconds: Double = 14

    /// Seconds a bubble stays. `base` is the time for a line of about eight
    /// words; longer lines get a little more, never past twice the base.
    public static func showTime(_ text: String, base: Double) -> Double {
        let words = Double(text.split(separator: " ").count)
        return min(base * 2, base / 2 + words * 0.9)
    }

    /// Replace every `{key}` in `template` with its value. Unknown keys are left as they are.
    public static func render(_ template: String, _ values: [String: String]) -> String {
        values.reduce(template) { text, pair in text.replacingOccurrences(of: "{\(pair.key)}", with: pair.value) }
    }

    /// One line, cleaned up the way a small model needs: first non-empty line,
    /// no wrapping quotes, no "Name:" prefix, no hidden-reasoning tags, capped.
    public static func cleanLine(_ raw: String, speaker: String, maxLength: Int = 160) -> String {
        var text = raw
        if let close = text.range(of: "</think>") { text = String(text[close.upperBound...]) }
        var line = text.split(whereSeparator: \.isNewline).map { $0.trimmingCharacters(in: .whitespaces) }
            .first { !$0.isEmpty } ?? ""
        for prefix in ["\(speaker):", "\(speaker.uppercased()):", "*\(speaker)*:"] where line.hasPrefix(prefix) {
            line = String(line.dropFirst(prefix.count)).trimmingCharacters(in: .whitespaces)
        }
        while let first = line.first, let last = line.last, line.count > 1,
              ["\"", "“", "'", "*"].contains(String(first)), ["\"", "”", "'", "*"].contains(String(last)) {
            line = String(line.dropFirst().dropLast()).trimmingCharacters(in: .whitespaces)
        }
        if line.count > maxLength {
            line = String(line.prefix(maxLength)).trimmingCharacters(in: .whitespaces) + "…"
        }
        return line
    }

    // MARK: What the model's marks mean on screen

    /// A stretch of a line with one look: `*sighs*` and `_really_` are italic,
    /// `**important**` bold, `***loud***` both.
    public struct StyledRun: Equatable, Sendable {
        public var text: String
        public var bold: Bool
        public var italic: Bool
        public init(_ text: String, bold: Bool = false, italic: Bool = false) {
            self.text = text; self.bold = bold; self.italic = italic
        }
    }

    /// The line cut into runs. A mark only counts when it opens against a
    /// non-space, closes after a non-space, and has its twin later in the line
    /// (`_` also needs a word boundary on the outside), so "2 * 3" and
    /// "snake_case" stay as written. Runs of spaces collapse to one.
    public static func styled(_ line: String) -> [StyledRun] {
        let chars = Array(line)
        var runs: [StyledRun] = []
        var buffer = ""
        var bold = false, italic = false
        var open: [(mark: Swift.Character, count: Int)] = []
        var lastWasSpace = false
        func flush() {
            if !buffer.isEmpty { runs.append(StyledRun(buffer, bold: bold, italic: italic)); buffer = "" }
        }
        func isWordChar(_ c: Swift.Character?) -> Bool { c.map { $0.isLetter || $0.isNumber } ?? false }
        func opens(at i: Int, _ n: Int) -> Bool {
            guard i + n < chars.count, !chars[i + n].isWhitespace, chars[i + n] != chars[i] else { return false }
            return chars[i] == "*" || !isWordChar(i > 0 ? chars[i - 1] : nil)
        }
        func closes(at i: Int, _ n: Int) -> Bool {
            guard i > 0, !chars[i - 1].isWhitespace, chars[i - 1] != chars[i] else { return false }
            return chars[i] == "*" || !isWordChar(i + n < chars.count ? chars[i + n] : nil)
        }
        func runLength(at i: Int) -> Int {
            var n = 1
            while i + n < chars.count, chars[i + n] == chars[i] { n += 1 }
            return n
        }
        func hasCloser(_ mark: Swift.Character, _ n: Int, from start: Int) -> Bool {
            var j = start
            while j < chars.count {
                if chars[j] == mark {
                    let m = runLength(at: j)
                    if m == n, closes(at: j, n) { return true }
                    j += m
                } else { j += 1 }
            }
            return false
        }
        func apply(_ mark: Swift.Character, _ n: Int, on: Bool) {
            switch n {
            case 1: italic = on
            case 2: bold = on
            default: bold = on; italic = on
            }
        }
        var i = 0
        while i < chars.count {
            let c = chars[i]
            if c == "*" || c == "_" {
                let n = runLength(at: i)
                if n <= 3, let top = open.last, top.mark == c, top.count == n, closes(at: i, n) {
                    flush(); apply(c, n, on: false); open.removeLast(); i += n; continue
                }
                if n <= 3, opens(at: i, n), hasCloser(c, n, from: i + n) {
                    flush(); apply(c, n, on: true); open.append((c, n)); i += n; continue
                }
                buffer += String(repeating: c, count: n); lastWasSpace = false; i += n; continue
            }
            if c == " " {
                if !lastWasSpace { buffer.append(c) }
                lastWasSpace = true
            } else {
                buffer.append(c); lastWasSpace = false
            }
            i += 1
        }
        flush()
        return runs
    }

    /// The line with the marks taken out, for anywhere that cannot show a style.
    public static func plain(_ line: String) -> String { styled(line).map(\.text).joined() }
}

File: win/LedgelingsCore/Banter.cs
namespace Ledgelings.Core;

/// <summary>Who a creature is when it opens its mouth.</summary>
public sealed record Character(string Name, string Persona);

/// <summary>The words that go to the model. Pure string work, so it is testable without
/// a server: templates with <c>{placeholders}</c>, filled from a small dictionary.</summary>
public static class Banter
{
    public static readonly IReadOnlyList<string> Placeholders = new[]
    {
        "speaker", "speakerKind", "speakerPersona", "listener", "listenerKind", "listenerPersona", "situation", "line",
    };

    /// <summary>What the built-in creature is, for the prompt.</summary>
    public const string DefaultKind = "a small square creature";

    public const string DefaultSystemPrompt =
        "You are {speaker}, {speakerKind}, living on the edge of a computer screen. {speakerPersona}\n" +
        "You are talking to {listener}, {listenerKind}, who lives on the same edge. {listenerPersona}\n" +
        "Say ONE line to {listener}: a joke, a jab or a tease, at most 20 words, in your own voice.\n" +
        "Output only the line. No quotes, no name prefix, no explanation.";

    public const string DefaultLinePrompt =
        "Right now: {situation}\n" +
        "Say your line to {listener}.";

    public const string DefaultReplyPrompt =
        "Right now: {situation}\n" +
        "{listener} just said to you: \"{line}\"\n" +
        "Answer back in ONE line, in character, at most 20 words.";

    public static readonly IReadOnlyList<Character> DefaultCharacters = new[]
    {
        new Character("Blocky", "Grumpy and proud. Hates the mouse cursor. Thinks the bottom edge is the only respectable edge."),
        new Character("Pip", "Cheerful and easily impressed. Loves the ceiling. Laughs at everything, including insults."),
        new Character("Mortimer", "Old and philosophical. Speaks slowly, quotes wisdom he made up, sighs a lot."),
        new Character("Zed", "Sleepy. Would rather be napping. Every sentence drifts toward bed."),
        new Character("Dot", "Tiny, fast and sarcastic. Brags about speed. Calls everyone else a boulder."),
        new Character("Ruth", "Bossy, organised, keeps count of everything. Disapproves of jumping."),
    };

    /// <summary>How long a bubble stays up for a line of typical length, in seconds.</summary>
    public const double DefaultBubbleSeconds = 14;

    /// <summary>Seconds a bubble stays. <paramref name="baseSeconds"/> is the time for a line of about eight
    /// words; longer lines get a little more, never past twice the base.</summary>
    public static double ShowTime(string text, double baseSeconds)
    {
        double words = text.Split(' ', StringSplitOptions.RemoveEmptyEntries).Length;
        return Math.Min(baseSeconds * 2, baseSeconds / 2 + words * 0.9);
    }

    /// <summary>Replace every <c>{key}</c> in <paramref name="template"/> with its value. Unknown keys are left as they are.</summary>
    public static string Render(string template, IReadOnlyDictionary<string, string> values)
    {
        var text = template;
        foreach (var (key, value) in values) text = text.Replace("{" + key + "}", value);
        return text;
    }

    /// <summary>One line, cleaned up the way a small model needs: first non-empty line,
    /// no wrapping quotes, no "Name:" prefix, no hidden-reasoning tags, capped.</summary>
    public static string CleanLine(string raw, string speaker, int maxLength = 160)
    {
        var text = raw;
        var close = text.IndexOf("</think>", StringComparison.Ordinal);
        if (close >= 0) text = text[(close + "</think>".Length)..];
        var line = text.Split('\n', '\r').Select(l => l.Trim(' ', '\t')).FirstOrDefault(l => l.Length > 0) ?? "";
        foreach (var prefix in new[] { speaker + ":", speaker.ToUpperInvariant() + ":", "*" + speaker + "*:" })
            if (line.StartsWith(prefix, StringComparison.Ordinal)) line = line[prefix.Length..].Trim(' ', '\t');
        var opens = new[] { '"', '\u201c', '\'', '*' };
        var closes = new[] { '"', '\u201d', '\'', '*' };
        while (line.Length > 1 && opens.Contains(line[0]) && closes.Contains(line[^1]))
            line = line[1..^1].Trim(' ', '\t');
        if (line.Length > maxLength) line = line[..maxLength].Trim(' ', '\t') + "\u2026";
        return line;
    }

    // MARK: What the model's marks mean on screen

    /// <summary>A stretch of a line with one look: <c>*sighs*</c> and <c>_really_</c> are italic,
    /// <c>**important**</c> bold, <c>***loud***</c> both.</summary>
    public readonly record struct StyledRun(string Text, bool Bold = false, bool Italic = false);

    /// <summary>The line cut into runs. A mark only counts when it opens against a non-space, closes
    /// after a non-space, and has its twin later in the line (<c>_</c> also needs a word boundary on
    /// the outside), so "2 * 3" and "snake_case" stay as written. Runs of spaces collapse to one.</summary>
    public static IReadOnlyList<StyledRun> Styled(string line)
    {
        var chars = line.ToCharArray();
        var runs = new List<StyledRun>();
        var buffer = new System.Text.StringBuilder();
        bool bold = false, italic = false, lastWasSpace = false;
        var open = new Stack<(char Mark, int Count)>();

        void Flush()
        {
            if (buffer.Length > 0) { runs.Add(new StyledRun(buffer.ToString(), bold, italic)); buffer.Clear(); }
        }
        static bool IsWordChar(char? c) => c is char ch && char.IsLetterOrDigit(ch);
        bool Opens(int i, int n)
        {
            if (i + n >= chars.Length || char.IsWhiteSpace(chars[i + n]) || chars[i + n] == chars[i]) return false;
            return chars[i] == '*' || !IsWordChar(i > 0 ? chars[i - 1] : null);
        }
        bool Closes(int i, int n)
        {
            if (i == 0 || char.IsWhiteSpace(chars[i - 1]) || chars[i - 1] == chars[i]) return false;
            return chars[i] == '*' || !IsWordChar(i + n < chars.Length ? chars[i + n] : null);
        }
        int RunLength(int i)
        {
            var n = 1;
            while (i + n < chars.Length && chars[i + n] == chars[i]) n++;
            return n;
        }
        bool HasCloser(char mark, int n, int start)
        {
            var j = start;
            while (j < chars.Length)
            {
                if (chars[j] == mark)
                {
                    var m = RunLength(j);
                    if (m == n && Closes(j, n)) return true;
                    j += m;
                }
                else j++;
            }
            return false;
        }
        void Apply(int n, bool on)
        {
            switch (n)
            {
                case 1: italic = on; break;
                case 2: bold = on; break;
                default: bold = on; italic = on; break;
            }
        }

        var i = 0;
        while (i < chars.Length)
        {
            var c = chars[i];
            if (c == '*' || c == '_')
            {
                var n = RunLength(i);
                if (n <= 3 && open.Count > 0 && open.Peek().Mark == c && open.Peek().Count == n && Closes(i, n))
                {
                    Flush(); Apply(n, false); open.Pop(); i += n; continue;
                }
                if (n <= 3 && Opens(i, n) && HasCloser(c, n, i + n))
                {
                    Flush(); Apply(n, true); open.Push((c, n)); i += n; continue;
                }
                buffer.Append(c, n); lastWasSpace = false; i += n; continue;
            }
            if (c == ' ')
            {
                if (!lastWasSpace) buffer.Append(c);
                lastWasSpace = true;
            }
            else
            {
                buffer.Append(c); lastWasSpace = false;
            }
            i++;
        }
        Flush();
        return runs;
    }

    /// <summary>The line with the marks taken out, for anywhere that cannot show a style.</summary>
    public static string Plain(string line) => string.Concat(Styled(line).Select(r => r.Text));
}
